pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeRangeSegmentPruner.java

```diff
@@ -0,0 +1,348 @@
+      return range;
+    }
+
+    long startTime = znRecord.getLongField(CommonConstants.Segment.START_TIME, -1);
+    long endTime = znRecord.getLongField(CommonConstants.Segment.END_TIME, -1);
+    if (startTime < 0 || endTime < 0 || startTime > endTime) {
+      LOGGER.warn("Failed to find valid end time for segment: {}, table: {}", segment, _tableNameWithType);
+      return range;
+    }
+
+    TimeUnit timeUnit = znRecord.getEnumField(CommonConstants.Segment.TIME_UNIT, TimeUnit.class, TimeUnit.DAYS);
+    range[0] = timeUnit.toMillis(startTime);
+    range[1] = timeUnit.toMillis(endTime);
+    return range;
+  }
+
+  @Override
+  public void refreshSegment(String segment) {
+    Set<String> segments;
+    if (_timeRangeMSToSegmentSearchTree == null) {
+      segments = new HashSet<>();
+    } else {
+      segments = _timeRangeMSToSegmentSearchTree.getAllValues();
+    }
+    if (!segment.contains(segment)) {
+      segments.add(segment);
+    }
+
+    updateTimeRangeMSToSegmentSearchTree(segments);
+  }
+
+  /**
+   * NOTE: Pruning is done by searching _timeRangeToSegmentSearchTree based on request time range and check if the results
+   *       are in the input segments. By doing so we will have run time O(M * logN) (N is the # of all online segments,
+   *       M is the # of qualified intersected segments).
+   */
+  @Override
+  public Set<String> prune(BrokerRequest brokerRequest, Set<String> segments) {
+    // The pruned order may be different from the input order
+    if (_timeRangeMSToSegmentSearchTree == null) {
+      return segments;
+    }
+    FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
+    if (filterQueryTree == null) {
+      return segments;
+    }
+    List<Interval> ranges = getFilterTimeRange(filterQueryTree);
+    if (ranges == null) { // cannot prune based on time for input request
+      return segments;
+    }
+    if (ranges.size() == 0) { // invalid query time range
+      return Collections.emptySet();
+    }
+
+    Set<String> prunedSegments = new HashSet<>();
+    for (Interval range : ranges) {
+      Interval rangeMS = convertRangeToMS(range);
+      for (Object segment : _timeRangeMSToSegmentSearchTree.searchAll(rangeMS)) {
+        if (segments.contains(segment)) {
+          prunedSegments.add((String) segment);
+        }
+      }
+    }
+    return new HashSet<String>(prunedSegments);
+  }
+
+  private List<Interval> getFilterTimeRange(FilterQueryTree filterQueryTree) {
+    // return NUll if no time range info or cannot filter base on the info (e.g. 'SELECT * from myTable where time < 50 OR firstName = Jason')
+    // return an empty list if filtering range is specified but invalid (e.g. 'SELECT * from myTable where time < 50 AND time > 100')
+    switch (filterQueryTree.getOperator()) {
+      case AND:
+        List<List<Interval>> andRanges = new ArrayList<>();
```

Comment:
I am not quite sure why list of list is needed here?
Each child operator under AND root will give you a list containing one or more intervals. 
This is List<Interval> childRanges.
You can still maintain the result list as List<Interval> resultRanges and do the following inside for loop

```
for (FilterQueryTree child : filterQueryTree.getChildren()) {
          List<Interval> childRanges = getFilterTimeRange(child);
          if (childRanges != null) {
            resultRanges.addAll(childRanges);
          }
        }
        // do sorted merge of these intervals
        return getIntersectionSortedRanges(resultIntervals);

```

Reply:
Say I have one child ranges as { [1, 4], [8, 10]} (from OR operator), another as {[3, 5], [7, 9]}. The merged result should be {[3, 4], [8, 9]}. If I put all child ranges into one list as { [1, 4], [8, 10], [3, 5], [7, 9]}, some helper informations are needed to track which ranges are from the same child to avoid calculate the intersections between them. In this case I should not calculate the intersection between [1, 4] and [8, 10].

Reply:
Sorted merge is an O(N) operation. So the algorithm will first sort { [1, 4], [8, 10], [3, 5], [7, 9]} on startTime to 
{ [1, 4], [3, 5], [7, 9], [8, 10]} and then do a single pass through them to merge into disjoint intervals resulting in {[3, 4], [8, 9]}. 

I don't think keeping track of which range came from which child is worth the code complexity. Time complexity is still O(N). Also, the time filters are usually not that deeply rooted in a complex filter tree. While, there could be multiple occurrences of time filter in a query (haven't seen many example so far though), they are generally not going to be part of every subtree filter expression.

Reply:
I see your points now, yes, this can be simplified to list by sorting in the function. The time complexity may not be big a concern since the FilterQuery tree should not be too complex. I'll refactor this.